### PR TITLE
Add NRules.Testing project

### LINKED ISSUE #103: Test library
Create a library (and publish it via nuget) to allow easy authoring of rule unit tests.
The library should only depend on NRules and not any particular test runner or assertion framework.

### PR DESCRIPTION
**Pull request by gritcsenko**
And use it in integration tests
Closes #103

### CODE REVIEW

**Comment by snikolayev** [comment]
@gritcsenko this also has conflicts now

File: src/NRules/NRules/Aggregators/AggregatorRegistry.cs [hunk 1]
```diff
@@ -3,10 +3,17 @@
 
 namespace NRules.Aggregators
 {
+    public interface IAggregatorRegistry
```
Comment (snikolayev) [inline src/NRules/NRules/Aggregators/AggregatorRegistry.cs]: @gritcsenko it looks like all/most changes in the NRules solution are unrelated to the extraction of the test fixture into its own project, so I would like to take them out of this PR

File: src/NRules/Tests/NRules.IntegrationTests/TestAssets/BaseRuleTestFixture.cs [hunk 2]
```diff
@@ -1,191 +1,23 @@
...
+﻿using NRules.Fluent;
+using NRules.Testing;
 
 namespace NRules.IntegrationTests.TestAssets
 {
     public abstract class BaseRuleTestFixture
     {
-        protected readonly ISession Session;
-        protected readonly RuleRepository Repository;
+        private readonly RuleTestFixture _fixture;
```
Comment (snikolayev) [inline src/NRules/Tests/NRules.IntegrationTests/TestAssets/BaseRuleTestFixture.cs]: @gritcsenko I do generally prefer composition over inheritance, but in this case, why shouldn't the tests directly inherit from the RuleTestFixture?

File: src/NRules/NRules/Rete/JoinNode.cs [hunk 3]
```diff
@@ -127,13 +127,13 @@ public override void PropagateUpdate(IExecutionContext context, List<Fact> facts
...
                 var joinedSets = JoinedSets(context, facts);
                 foreach (var set in joinedSets)
-                foreach (var fact in set.Facts)
-                {
-                    if (MatchesConditions(context, set.Tuple, fact))
-                        toUpdate.Add(set.Tuple, fact);
-                    else
-                        toRetract.Add(set.Tuple, fact);
-                }
+                    foreach (var fact in set.Facts)
```
Comment (snikolayev) [inline src/NRules/NRules/Rete/JoinNode.cs]: @gritcsenko it's somewhat annoying that these code styling changes are sprinkled throughout the PR. Makes it very hard to see real changes. Can you push all code styling changes as a single PR on top of what's currently in develop? Then rebase the rest of PRs once those style changes are accepted and merged.

Comment (snikolayev) [reply src/NRules/NRules/Rete/JoinNode.cs]: What are you using for linting the code?

**Review by snikolayev — CHANGES_REQUESTED** [review]

File: src/NRules/NRules/Rete/JoinNode.cs (hunk 3 above)
Comment (gritcsenko) [reply src/NRules/NRules/Rete/JoinNode.cs]: > What are you using for linting the code?

Visual Studio 2022 v 17.4.1 
No extensions (bare)

File: src/NRules/NRules/Aggregators/AggregatorRegistry.cs (hunk 1 above)
Comment (gritcsenko) [reply src/NRules/NRules/Aggregators/AggregatorRegistry.cs]: I've reverted all the changes in NRules project. Sorry

File: src/NRules/NRules/Rete/JoinNode.cs (hunk 3 above)
Comment (gritcsenko) [reply src/NRules/NRules/Rete/JoinNode.cs]: I've reverted all the changes in NRules project. Sorry

File: src/NRules/Tests/NRules.IntegrationTests/TestAssets/BaseRuleTestFixture.cs (hunk 2 above)
Comment (gritcsenko) [reply src/NRules/Tests/NRules.IntegrationTests/TestAssets/BaseRuleTestFixture.cs]: Let me try it

Comment (gritcsenko) [reply src/NRules/Tests/NRules.IntegrationTests/TestAssets/BaseRuleTestFixture.cs]: @snikolayev I've implemented inheritance. Please, take a look

**Comment by gritcsenko** [comment]
TODO for me:
[x] Update build scripts (Done :heavy_check_mark:)

File: src/NRules/Tests/NRules.IntegrationTests/TestAssets/XUnitRuleAsseter.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using NRules.Testing;
+
+namespace NRules.IntegrationTests.TestAssets;
+
+public sealed class XUnitRuleAsseter : IRuleAsseter
```
Comment (snikolayev) [inline src/NRules/Tests/NRules.IntegrationTests/TestAssets/XUnitRuleAsseter.cs]: @gritcsenko typo - Asserter

File: src/NRules/NRules.Testing/IRepositorySetup.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿using NRules.Fluent;
+using NRules.Fluent.Dsl;
+
+namespace NRules.Testing;
+
+public interface IRepositorySetup
+{
+    void Rule<T>() where T : Rule;
+
+    void Rule(IRuleMetadata metadata);
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/IRepositorySetup.cs]: @gritcsenko I don't think you need an overload with IRuleMetadata. You just need a non-generic Rule(Type ruleType) and instantiate the metadata inside.

File: src/NRules/NRules.Testing/IRuleFireAssertion.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using NRules.Fluent.Dsl;
+
+namespace NRules.Testing;
+
+public interface IRuleFireAssertion
+{
+    IRuleFireAssertResult IsFiredTimes(int expected);
+
+    IRuleFireAssertResult IsFiredTimes<T>(int expected) where T : Rule;
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/IRuleFireAssertion.cs]: @gritcsenko probably better to say HasFiredTimes from the English readability perspective

File: src/NRules/NRules.Testing/IRuleTestFixture.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections.Generic;
+using NRules.Fluent.Dsl;
+
+namespace NRules.Testing;
+
+public interface IRuleTestFixture
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/IRuleTestFixture.cs]: @gritcsenko maybe better to name IRulesTestFixture given it can have multiple rules in the session?

File: src/NRules/NRules.Testing/ISelectedRuleVerification.cs [hunk 8]
```diff
@@ -0,0 +1,6 @@
+﻿namespace NRules.Testing;
+
+public interface ISelectedRuleVerification
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/ISelectedRuleVerification.cs]: @gritcsenko not sure if "selected" adds anything here, vs just naming IRuleVerification. What do you think?

File: src/NRules/NRules.Testing/RuleTestFixture.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+namespace NRules.Testing;
+
+public class RuleTestFixture : IRuleTestFixture
+{
+    private readonly Lazy<ISession> _lazySession;
+    private readonly IRuleActivator _ruleActivator;
+    private readonly RuleCompiler _compiler;
+    private readonly RepositorySetup _setup;
+
+    public RuleTestFixture(IRuleActivator ruleActivator, RuleCompiler compiler, IRuleAsseter asseter)
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/RuleTestFixture.cs]: @gritcsenko typo in iface name and parameter name: asserter

File: src/NRules/NRules.Testing/RuleTestFixture.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            Activator = ruleActivator
+        };
+        _setup = new RepositorySetup(repository);
+        Verify = new RulesVerification(asseter, _setup);
+
+        _lazySession = new(CreateSession);
+
+        ISession CreateSession()
+        {
+            var ruleDefinitions = repository.GetRuleSets().SelectMany(set => set.Rules);
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/RuleTestFixture.cs]: @gritcsenko there is an extension method on IRuleRepository - GetRules

File: src/NRules/NRules.Testing/CachedRuleActivator.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NRules.Fluent;
+using NRules.Fluent.Dsl;
+
+namespace NRules.Testing;
+
+public class CachedRuleActivator : IRuleActivator
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/CachedRuleActivator.cs]: @gritcsenko we'll need XML docs on public types, but we can do it separately, after we settle on the code change (can even be a separate PR)

File: src/NRules/NRules.Testing/CachedRuleActivator.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NRules.Fluent;
+using NRules.Fluent.Dsl;
+
+namespace NRules.Testing;
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/CachedRuleActivator.cs:7]: @gritcsenko I'm still internally debating between NRules.Test and NRules.Testing. I'll think a bit more. Do you think NRules.Test will cause name clashes?

File: src/NRules/NRules.sln [hunk 13]
```diff
@@ -19,6 +19,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NRules.Json", "NRules.Json\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NRules.Json.Tests", "Tests\NRules.Json.Tests\NRules.Json.Tests.csproj", "{846AAC54-E7D1-4167-8D53-13B836DE7CFA}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NRules.Testing", "NRules.Testing\NRules.Testing.csproj", "{C35D16EE-6FFD-47B4-91F9-0930A0CCD699}"
```
Comment (snikolayev) [inline src/NRules/NRules.sln:22]: @gritcsenko we'll need to create a nuspec file and integrate into build/documentation. But I can easily do it after the PR is merged

File: src/NRules/NRules.Testing/CachedRuleActivator.cs (hunk 12 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/CachedRuleActivator.cs:7]: @snikolayev I think that namespaces should be named as groups or collections, not as a single entitiy to avoid name clashes with class names.

File: src/NRules/Tests/NRules.IntegrationTests/TestAssets/XUnitRuleAsseter.cs (hunk 4 above)
Comment (gritcsenko) [reply src/NRules/Tests/NRules.IntegrationTests/TestAssets/XUnitRuleAsseter.cs]: Fixed, thanks

File: src/NRules/NRules.Testing/IRepositorySetup.cs (hunk 5 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/IRepositorySetup.cs]: Updated, thanks

File: src/NRules/NRules.Testing/IRuleFireAssertion.cs (hunk 6 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/IRuleFireAssertion.cs]: This interface was leftover. Removed

File: src/NRules/NRules.Testing/IRuleTestFixture.cs (hunk 7 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/IRuleTestFixture.cs]: Renamed, thanks

File: src/NRules/NRules.Testing/ISelectedRuleVerification.cs (hunk 8 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/ISelectedRuleVerification.cs]: Renamed, thanks

File: src/NRules/NRules.Testing/RuleTestFixture.cs (hunk 9 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/RuleTestFixture.cs]: Fixed, thanks

File: src/NRules/NRules.Testing/RuleTestFixture.cs (hunk 10 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/RuleTestFixture.cs]: Now I know. Thanks

File: src/NRules/NRules.Testing/CachedRuleActivator.cs (hunk 11 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/CachedRuleActivator.cs]: Added XML Docs to all public types and members

File: src/NRules/NRules.sln (hunk 13 above)
Comment (gritcsenko) [reply src/NRules/NRules.sln:22]: .nuspec was added. Thanks for the remider

File: build.ps1 [hunk 14]
```diff
@@ -10,34 +10,36 @@ if (Test-Path Env:CI) { $version = $Env:APPVEYOR_BUILD_VERSION }
 if (Test-Path Env:CI) { $configuration = $Env:CONFIGURATION }
 
 $components = @{
-    'NRules' = @{
-        name = 'NRules'
+    'NRules'                     = @{
```
Comment (snikolayev) [inline build.ps1]: @gritcsenko what's up with these alignment changes? Is this also built-in VS linting?

File: src/NRules/NRules.Testing/IRuleVerification.cs [hunk 15]
```diff
@@ -1,6 +1,13 @@
 ﻿namespace NRules.Testing;
 
+/// <summary>
+/// Represents specific rule verification
+/// </summary>
 public interface IRuleVerification
 {
+    /// <summary>
+    /// Asserts that rule was fired exact amount of times
```
Comment (snikolayev) [inline src/NRules/NRules.Testing/IRuleVerification.cs]: @gritcsenko ".. rule fired exact number of times".
The rule is not fired by someone else, but it itself fires (at least that the terminology I saw in the industry for inference engines).

File: src/NRules/NRules.Testing/CachedRuleActivator.cs (hunk 12 above)
Comment (snikolayev) [reply src/NRules/NRules.Testing/CachedRuleActivator.cs:7]: @gritcsenko that's an interesting observation. I also think there are plenty of examples where it's in singular, when referring to a concept, or if it's a namespace that matches the package name. But I think then "Testing" nicely fits that idea of representing a concept.

File: build.ps1 (hunk 14 above)
Comment (gritcsenko) [reply build.ps1]: This is VS Code's PowerShell plugin messing up. I'll revert.

File: src/NRules/NRules.Testing/IRuleVerification.cs (hunk 15 above)
Comment (gritcsenko) [reply src/NRules/NRules.Testing/IRuleVerification.cs]: Fixed

File: build.ps1 (hunk 14 above)
Comment (gritcsenko) [reply build.ps1]: Fixed

**Review by snikolayev — APPROVED** [review]
